soc/xtensa/intel_adsp/tools/cavstool.py

```diff
@@ -659,30 +660,65 @@ async def _main(server):
         if dsp.HIPCIDA & 0x80000000:
             dsp.HIPCIDA = 1<<31 # must ACK any DONE interrupts that arrive!
 
+        if server:
+            # Check if the client connection is alive.
+            if not is_connection_alive(server):
+                lock.acquire()
+                start_output = False
+                lock.release()
+
 class adsp_request_handler(socketserver.BaseRequestHandler):
     """
     The request handler class for control the actions of server.
     """
 
-    def receive_fw(self, filename):
+    def receive_fw(self):
+        header_sz = 80
+
+        log.info("Receiving...")
+        # The first contain the header and the front
+        # data of the firmware
+        d = self.request.recv(BUF_SIZE)
+
+        # Unpacked the header data
+        # Include size(4), filename(42), MD5(32) and pad(2)
+        header = d[:header_sz]
+        total = d[header_sz:]
+        cnt = len(total)
+        s = struct.Struct('I 42s 32s H')
+        fsize, fname, md5_tx_b, pad = s.unpack(header)
+        log.info(f'size:{fsize}, filename:{fname}, MD5:{md5_tx_b}, pad:{pad})')
+
+
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while cnt < fsize:
+            data = self.request.recv(BUF_SIZE)
+            if not data:
+                break
+
+            total += data
```

Comment:
Just use a very large BUF_SIZE (16MB) and read once. Keep it simple. If you look at the size of the files in https://github.com/thesofproject/sof-bin/tree/main/v2.2.x/sof-v2.2-rc1/intel-signed you can see that the largest file is `sof-cnl.ri` and it's 545KB big and it's 96% full.

The ADSP memory is many, many orders of magnitude smaller compared to the systems this script runs on.
```
sof/tools/sof_ri_info/sof_ri_info.py v2.2.x/sof-v2.2-rc1/sof-cnl.ri 

  l2 hpsram                           0xbe000000 (3031040 + 114688  96.35% used)
```

Reply:
Please kindly correct me if I misunderstand your suggestion.  I tried this but it seems like it doesn't work well with large BUF_SIZE,.
```
    def receive_fw(self):
        header_sz = 80
        LARGE_BUF_SIZE = 16 * 1024 * 1024

        log.info("Receiving...")
        # The first contain the header and the front
        # data of the firmware
        d = self.request.recv(BUF_SIZE)

        # Unpacked the header data
        # Include size(4), filename(42), MD5(32) and fnamelen(2)
        header = d[:header_sz]
        total = d[header_sz:]
        s = struct.Struct(PACKET_HEADER_FORMAT_FW)
        fsize, fname, md5_tx_b, fnamelen = s.unpack(header)
        log.info(f'size:{fsize}, filename:{fname}, MD5:{md5_tx_b}, fnamelen:{fnamelen})')

        #receive all once in a large buffer.
        total += self.request.recv(LARGE_BUF_SIZE)  

        cnt = len(total)
        log.info(f"Done Receiving {cnt}.")
```
And it got 1368 most in our network environment.
```
INFO:cavs-fw:Req server start...
INFO:cavs-fw:Log server start...
INFO:cavs-fw:192.168.0.200 wrote: b'download'
INFO:cavs-fw:Receiving...
INFO:cavs-fw:size:114704, filename:b'zephyr.ri.9c23b0cc8538e7a58b9246a0b77e9b59', MD5:b'bbfe7524c87071c99fa737b4946a0ef9', fnamelen:42)
INFO:cavs-fw:Done Receiving 1368.
ERROR:cavs-fw:MD5 mismatch: bbfe7524c87071c99fa737b4946a0ef9 vs. 3d6f257012bf0f2a93428208b5d75a1a
ERROR:cavs-fw:Receive firmware failed.

```
So I suggest to keep the current way with enlarging the buffer size.

Reply:
Apparently we can't have nice things... there's apparently nothing like `recvall()` in the Python socket API, so you're right: the `while recv()` dance is required. Sorry for misleading you here.

Reply:
not at all, thank you so much for this suggestion!